### DISCUSSION TITLE
Is there any function like computeIfAbsent in Java

### DISCUSSION THREAD

**Discussion by tasty0tomato** [dotnet/runtime#67376] [Q&A] [answered]
https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html#computeIfAbsent-K-java.util.function.Function-
It seems easy to use this function to do Add operation like this
` buckets.computeIfAbsent(word.length(), x -> new ArrayList()).add(word);`
But in C#, and maybe because I am not much familiar with C#, I need to write these code.
```        
if (!buckets.ContainsKey(word.Length))
{
   buckets.Add(word.Length, new List<string>());
}
buckets[word.Length].Add(word);
```
And there was a similar question in [Stack](https://stackoverflow.com/questions/68964120/whats-the-quivalent-of-java-maps-computeifabsent-in-c-sharp-dictionary), but unfortunately, there wasn't an answer.

**Comment by gfoidl** [comment, marked as answer]
You can write it in C# with very efficiently. See the following example:
```c#
using System.Runtime.InteropServices;

Dictionary<int, List<string>> buckets = new();

AddToBucket(buckets, "Hello");
AddToBucket(buckets, "World");
AddToBucket(buckets, "C# rocks :-)");

foreach (KeyValuePair<int, List<string>> bucket in buckets.OrderBy(b => b.Key))
{
    Console.WriteLine($"bucket: {bucket.Key}");

    foreach (string item in bucket.Value)
    {
        Console.WriteLine($"\t{item}");
    }
}

static void AddToBucket(Dictionary<int, List<string>> buckets, string word)
{
    ref List<string>? list = ref CollectionsMarshal.GetValueRefOrAddDefault(buckets, word.Length, out _);
    list ??= new List<string>();

    list.Add(word);
}
```
very efficiently.

---

Different approach:
```c#
static void AddToBucket(Dictionary<int, List<string>> buckets, string word)
{
    List<string> entries = buckets.GetValueOrDefault(word.Length, new List<string>());

    entries.Add(word);
    buckets[word.Length] = entries;
}
```

**Reply by tasty0tomato** [reply]
Wow, I do not even know _CollectionsMarshal_. And what a pity there isn't much information about it except _.NET API browser_.

**Reply by Joe4evr** [reply]
Yeah, `CollectionsMarshal` is pretty new.
